### ISSUE TITLE
lesspipe becomes <defunct> when using less +F

### ISSUE THREAD

**Issue by wilriker** [wofr06/lesspipe.sh#4]
Hello

I had an annoying issue for a long time where I was unable to follow log files in less with `+F` option (`--follow-name` was working though). It just did not update the content after starting less.

Today I finally found out that this has something to do with lesspipe and especially code2color. My lesspipe is set up to always try to do code highlighting and code2color is also present. So it pumps all the content through code2color. Using `pstree` just after opening the file (though without `+F` option) produces the following (sub)tree

```
|-zsh(9620)───less(9788)───lesspipe.sh(9789)───perl(9801)
```

where the `perl` process is code2color being executed. When I now press `F` (or use `+F` option when starting less) to go into follow mode the `perl` process will be removed from the above tree and the output of `ps` for the lesspipe.sh process changes to

```
 9789  0.0  0.0      0     0 pts/4    Z+   16:12   0:00 [lesspipe.sh] <defunct>
```

I encountered this behavior on CentOS 6.7 machine running perl 5.10.1 and bash 4.1.2 as well as on an Arch Linux running perl 5.22.1 and bash 4.3.042.

When I simply rename code2color to something else (thereby skipping code2color invocation) everything works as expected, so I assume this has something to do with code2color and/or perl.

Sadly, I have no experience in perl and so I cannot give you any more details about this.

Is that something that can be fixed or is this a known limitation when using this combination of less + bash + perl?

Best regards
Manuel

**Comment by wilriker** [comment]
I researched this a little more. lesspipe goes into `defunct` state immediately when the input file is less than 80 KiB in size (I guess that is some kind of buffer size in perl/code2color that keeps the perl process alive if the input is larger and therefore lesspipe also stays alive).

**Comment by wofr06** [comment]
On Wed, 13 Jan 2016, Manuel wrote:

> I researched this a little more. lesspipe goes into defunct state immediately when the input file is less than 80 KiB in size (I guess that is some kind of buffer
> size in perl/code2color that keeps the perl process alive if the input is larger and therefore lesspipe also stays alive).

the way how lesspipe is invoking external programs is not compatible with 
waiting for more input. The external programs are forked and when the end 
of data (end of file) is seen then the external program terminates. That 
should not only happen with colorful output (code2color) but with any 
interpreted contents, i.e. files that are not regarded as plain text.

That means, F within less or invoking less with +F will in most cases not 
work. I do not see any workaround except temporarily unsetting LESSOPEN to 
not trigger the use of lesspipe.sh.

Even with large files you will not see data appended after less was 
started, when lesspipe.sh is forking external programs.

## 

Best regards
Wolfgang Friebel

**Comment by wilriker** [comment]
That's basically what I already assumed that this is a (known) limitation.

I think usually you will only follow something like (plain-text) logfiles as I don't see that there is a case where you would for example follow a zip file being created or updated (I know, just because I don't see a case does not mean there isn't one). So what I mean by that is:

Is there a sensible way for lesspipe to detect either startup with `+F` or follow mode being triggered by `F`? If so, could it then just pass-through all data unmodified?

Best regards
Manuel

**Comment by wofr06** [comment]
On Thu, 14 Jan 2016, Manuel wrote:

> That's basically what I already assumed that this is a (known) limitation.
> 
> I think usually you will only follow something like (plain-text) logfiles as I don't see that there is a case where you would for example follow a zip file being
> created or updated (I know, just because I don't see a case does not mean there isn't one). So what I mean by that is:
> 
> Is there a sensible way for lesspipe to detect either startup with +F or follow mode being triggered by F? If so, could it then just pass-through all data unmodified?

This is indeed possible. Currently if the file command output 
matches ' script' or 'text executable' or does not match any of the many 
strings that decide what to do with the file then the string 'plain text' 
is set as file type. This triggers the call of code2color. If one would 
differentiate e.g. between 'plain text' and 'script text' depending on the 
file type, then code2color could be called conditionally only for 'script 
text'.

The difficulty is in the file command. In many cases its output is 
misleading and this could trigger the wrong behaviour. I have tested with 
a short text of mine containing just german sentences and some chars for 
structuring the text. The result is

my_file: UTF-8 Unicode C++ program text

However for most of the log files I get the strings

'ASCII text' or 'UTF-8 Unicode text'

These are probably the main cases for using +F and could be treated like 
you propose. I will see whether I can implement something along these 
lines.

Best regards
Wolfgang

**Comment by wilriker** [comment]
I also just tested tested a log file with `file` command and it told me

```
webapp.log: UTF-8 Unicode text, with very long lines
```

Maybe that amendment could also be considered as (hopefully) you don't have very long lines in source code. ;-) Though you actually might have very long lines in `.tex` files.

So, I think this comes down to extending the check in `isfinal` functions's

``` bash
if [[ "$1" = *plain\ text* ]]; then
    ...
fi
```

(near the end of the function) not only for the presence of code2color but also maybe for file endings like `*.log`, `*.log.*`, `*.txt` and others commonly known to contain non-source-code-plain-text and simply do nothing (using `cat` also leads to breaking). I guess this check needs to be done also at the top of `isfinal` when the separator is found.

Best regards
Manuel

**Comment by Felixoid** [comment]
Hello, dear @wofr06 and @wilriker.

It's quite a lot of time passed since the issue was updated last time. But it's still here, and I spent some time just recognizing the reason for misbehaving `less`.

Can it be somehow escalated?

**Comment by wofr06** [comment]
Ok, one could test for the presence of the +F when calling less. Then forking external functions could be suppressed. This could be achieved by a simple code snippet early on in the filetype function (just before the brotli test) such as
if [[ `ps -q $$ -o args=` = *+F* ]]; then
    return "plain text"
fi
As I do not have a test case right now, could you try the modification? If it works then I will  add it to lesspipe

**Comment by wofr06** [comment]
It is probably even better to put the test at the beginning of isfinal
if [[ ps -q $$ -o args= = +F ]]; then
return
fi

**Comment by wilriker** [comment]
This will though only solve the case where you actually start less with +F. For cases where you only go into following mode after starting less this won't help. Not complaining just clarifying.

**Comment by wofr06** [comment]
For that case I had in mind to call less with a second argument, an isolated colon. That argument could be saved early in the script and checked later similarly in isfinal. It is somewhat artificial but would solve the issue. This would also help to display a file as is if lesspipe errornously chose an unwanted transformation. What do you think?

**Comment by wilriker** [comment]
TBH I quit using lesspipe due to this limitation years ago because my main usage of less actually was to open a log file, go to the end, check if the expected line is already in there and if not start following until I saw the line I was interested in. What I want to say by that is: to me this is no longer important and my opinion should no longer have a weight with this.

**Comment by Felixoid** [comment]
I've tried it in `isfinal`, and it didn't help

```
isfinal() {
  typeset t
  if [[ $(ps -q $$ -o args=) = +F ]]; then
    return
  fi
  if [[ $3 = $sep$sep ]]; then
```

Neither in filetype

```
filetype () {
  # wrapper for 'file' command
  typeset name
  name="$1"
  if [[ "$1" = - ]]; then
    name="$filen"
  fi
  if [[ $(ps -q $$ -o args=) = +F ]]; then
    echo "plain text"
    return
  fi
  if [[ ("$name" = *.br || "$name" = *.bro || "$name" = *.tbr) ]]; then
```

Less doesn't follow the new lines in the file

The version header is: `lesspipe.sh, a preprocessor for less (version 1.83)`

**Comment by wofr06** [comment]
For some reasons my wildcard chars in the lines above disappeared. Of course it does not work without it. The correct test is
if [[ `ps -q $$ -o args=` = *+F* ]]; then
    ,,,
fi

**Comment by Felixoid** [comment]
Looks like it's wrong again. Can you, maybe, attach a file, please?

I've tried it like the following snippet, doesn't work =\

```
  if [[ $(ps -q $$ -o args=) = *+F* ]]; then
    echo "plain text"
    return
  fi
```

**Comment by wofr06** [comment]
No, it will not work. I just tried it myself by putting the test at the start of lesspipe. Then I read the less manpage (again).
There is an extremly simple solution to the issue:
less -L +F filename
As this solves the request I regard this problem solved

**Comment by Felixoid** [comment]
And it even works quite well for `F` in runtime!

Thank you for finding it.

**Comment by wilriker** [comment]
Just for the sake of completeness `less -L` will just skip lesspipe completely and open less without it. So it's not a general solution.

**Comment by wofr06** [comment]
I researched it a little further. The PID ($$) is that of the lesspipe call. The PPID is still not the less call but the shell invocation of lesspipe. Only the ppid of that process yields (for me) the process id that can be checked against +F
So the following code snippet finally works for me:
```bash
a=`ps -q $$ -o ppid=`
b=`ps -q $a -o ppid=`
if [[ `ps -q $b -o args=` = *+F* ]]; then
    exit 
fi
```

Putting it just at the start of the lesspipe.sh is equivalent to calling less with -L. To make it more useful it should be made conditionally on the file type. But I don't see the general consensus for what file types this should happen.

So I do not see an advantage of using this construct over using less -L

**Comment by wofr06** [comment]
Sorry, the copy paste removed again the wildcard stars and backticks
a=`ps -q $$ -o ppid=`
b=`ps -q $a -o ppid=`
if [[ `ps -q $b -o args=` = *+F* ]]; then
exit
fi

**Comment by wofr06** [comment]
Did not help: backtick before each ps and after the ppid= and args= strings, stars around +F

**Comment by Felixoid** [comment]
Most probably, this will help https://guides.github.com/features/mastering-markdown/

tl;dr: put an empty line, then three backticks only start the code block, your code, then close the block with three backticks again.

But you are right. The solution should be more general for all files that aren't supposed to open any preprocessor and therefore may be followed.

**Comment by wofr06** [comment]
lesspipe.sh checks now for a +F on the command line or a single colon as the last argument to the less call. In these cases lesspipe does nothing and watching growing files is possible
